client/attest.go

```diff
@@ -109,6 +129,76 @@ func (k *Key) getCertificateChain(client *http.Client) ([][]byte, error) {
 	return nil, fmt.Errorf("max certificate chain length (%v) exceeded", maxCertChainLength)
 }
 
+// SevSnpDevice encapsulates the SEV-SNP attestation device to add its attestation report
+// to a pb.Attestation.
+type SevSnpDevice struct {
+	Device sg.Device
+}
+
+// CreateSevSnpDevice opens the SEV-SNP attestation driver and wraps it with behavior
+// that allows it to add an attestation report to pb.Attestation.
+func CreateSevSnpDevice() (*SevSnpDevice, error) {
+	d, err := sg.OpenDevice()
+	if err != nil {
+		return nil, err
+	}
+	return &SevSnpDevice{Device: d}, nil
+}
+
+// AddAttestation will get the SEV-SNP attestation report given opts.TEENonce with
+// associated certificates and add them to `attestation`. If opts.TEENonce is empty,
+// then uses contents of opts.Nonce.
+func (d *SevSnpDevice) AddAttestation(attestation *pb.Attestation, opts AttestOpts) error {
+	var snpNonce [64]byte
+	if len(opts.TEENonce) == 0 {
+		copy(snpNonce[:], opts.Nonce[:])
+	} else if len(opts.TEENonce) != 64 {
+		return fmt.Errorf("the TEENonce size is %d. SEV-SNP device requires 64", len(opts.TEENonce))
+	} else {
+		copy(snpNonce[:], opts.TEENonce)
+	}
+	extReport, err := sg.GetExtendedReport(d.Device, snpNonce)
+	if err != nil {
+		return err
+	}
+	attestation.TeeAttestation = &pb.Attestation_SevSnpAttestation{
+		SevSnpAttestation: extReport,
+	}
+	return nil
+}
+
+// Close will free the device handle held by the SevSnpDevice. Calling more
+// than once has no effect.
+func (d *SevSnpDevice) Close() error {
+	if d.Device != nil {
+		err := d.Device.Close()
+		d.Device = nil
+		return err
+	}
+	return nil
+}
+
+// Does best effort to get a TEE hardware rooted attestation, but won't fail fatally
+// unless the user provided a TEEDevice object.
+func getTEEAttestationReport(attestation *pb.Attestation, opts AttestOpts) error {
+	device := opts.TEEDevice
+	if device != nil {
+		return device.AddAttestation(attestation, opts)
+	}
+
+	// Try SEV-SNP.
+	if device, err := CreateSevSnpDevice(); err == nil {
+		// Don't return errors if the attestation collection fails, since
+		// the user didn't specify a TEEDevice.
+		device.AddAttestation(attestation, opts)
```

Comment:
I wonder if we should log this at some point, otherwise debugging is impossible. The default go log package logs to stderr, which seems like a reasonable default. I'd like to avoid having users pass in a Logger opt. Thoughts?

Reply:
If you'd like "log", I can add it. I just didn't want to make log noise for folks not expecting the client to look for confidential compute capabilities.

Reply:
No, you're right. I think if we end up having issues with this, we can add a Logger opt that when nil defaults to not logging. Thanks!